File: CommonLib/BaseClassLib/src/main/java/com/yc/baseclasslib/manager/ProxyActivityListener.java
package com.yc.baseclasslib.manager;

import android.app.Activity;
import android.os.Bundle;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * <pre>
 *     @author yangchong
 *     blog  : https://github.com/yangchong211
 *     time  : 2018/11/9
 *     desc  : 代理类
 *     revise:
 * </pre>
 */
public class ProxyActivityListener extends ActivityLifecycleListener {

    private final Map<Class<Activity>, List<ActivityLifecycleListener>> mActivityLifecycleListeners = new HashMap<>();
    private final ActivityManager mActivityManager;

    ProxyActivityListener(ActivityManager manager) {
        this.mActivityManager = manager;
    }

    public void registerActivityLifecycleListener(Class<Activity> clazz, ActivityLifecycleListener lifecycleListener) {
        if (clazz == null || lifecycleListener == null) {
            return;
        }
        List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(clazz);
        if (lifecycleListeners == null) {
            lifecycleListeners = new ArrayList<>();
        }
        lifecycleListeners.add(lifecycleListener);
        mActivityLifecycleListeners.put(clazz, lifecycleListeners);
    }

    public boolean unregisterActivityLifecycleListener(Class<Activity> clazz, ActivityLifecycleListener lifecycleListener) {
        if (clazz == null || lifecycleListener == null) {
            return false;
        }
        synchronized (mActivityLifecycleListeners) {
            List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(clazz);
            if (lifecycleListeners == null) {
                return false;
            }
            boolean result = lifecycleListeners.remove(lifecycleListener);
            if (lifecycleListeners.size() == 0) {
                mActivityLifecycleListeners.remove(clazz);
            }
            return result;
        }
    }

    @Override
    public void onActivityResumed(Activity activity) {
        List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(activity.getClass());
        if (lifecycleListeners == null || lifecycleListeners.size() == 0) {
            return;
        }

        List<ActivityLifecycleListener> callbacks = new ArrayList<>(lifecycleListeners);
        for (int i = 0; i < callbacks.size(); i++) {
            callbacks.get(i).onActivityResumed(activity);
        }
    }

    @Override
    public void onActivityStopped(Activity activity) {
        List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(activity.getClass());
        if (lifecycleListeners == null || lifecycleListeners.size() == 0) {
            return;
        }

        List<ActivityLifecycleListener> callbacks = new ArrayList<>(lifecycleListeners);
        for (int i = 0; i < callbacks.size(); i++) {
            callbacks.get(i).onActivityStopped(activity);
        }
    }

    @Override
    public void onActivityPaused(Activity activity) {
        List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(activity.getClass());
        if (lifecycleListeners == null || lifecycleListeners.size() == 0) {
            return;
        }

        List<ActivityLifecycleListener> callbacks = new ArrayList<>(lifecycleListeners);
        for (int i = 0; i < callbacks.size(); i++) {
            callbacks.get(i).onActivityPaused(activity);
        }
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
        List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(activity.getClass());
        if (lifecycleListeners == null || lifecycleListeners.size() == 0) {
            return;
        }

        List<ActivityLifecycleListener> callbacks = new ArrayList<>(lifecycleListeners);
        for (int i = 0; i < callbacks.size(); i++) {
            callbacks.get(i).onActivitySaveInstanceState(activity, outState);
        }
    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        mActivityManager.remove(activity);
        List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(activity.getClass());
        if (lifecycleListeners == null || lifecycleListeners.size() == 0) {
            return;
        }

        List<ActivityLifecycleListener> callbacks = new ArrayList<>(lifecycleListeners);
        for (int i = 0; i < callbacks.size(); i++) {
            callbacks.get(i).onActivityDestroyed(activity);
        }
    }

    @Override
    public void onActivityStarted(Activity activity) {
        List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(activity.getClass());
        if (lifecycleListeners == null || lifecycleListeners.size() == 0) {
            return;
        }

        List<ActivityLifecycleListener> callbacks = new ArrayList<>(lifecycleListeners);
        for (int i = 0; i < callbacks.size(); i++) {
            callbacks.get(i).onActivityStarted(activity);
        }
    }

    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        mActivityManager.add(activity);
        List<ActivityLifecycleListener> lifecycleListeners = mActivityLifecycleListeners.get(activity.getClass());
        if (lifecycleListeners == null || lifecycleListeners.size() == 0) {
            return;
        }

        List<ActivityLifecycleListener> callbacks = new ArrayList<>(lifecycleListeners);
        for (int i = 0; i < callbacks.size(); i++) {
            callbacks.get(i).onActivityCreated(activity, savedInstanceState);
        }
    }
}

File: CommonLib/FragmentManager/src/main/java/com/yc/fragmentmanager/ProxyActivityListener.kt
package com.yc.fragmentmanager

import android.content.Context
import android.os.Bundle
import android.util.Log
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import java.util.ArrayList
import java.util.HashMap

/**
 * @author yangchong
 * blog  : https://github.com/yangchong211
 * GitHub : https://github.com/yangchong211/YCCommonLib
 * time  : 2018/11/9
 * desc  : 代理类
 * revise:
 */
class ProxyActivityListener : FragmentLifecycleListener() {

    private val mActivityLifecycleListeners: MutableMap<FragmentActivity, MutableList<FragmentLifecycleListener>> =
        HashMap()

    fun registerFragmentLifecycleListener(
        activity: FragmentActivity?,
        lifecycleListener: FragmentLifecycleListener?
    ) {
        if (activity == null || lifecycleListener == null) {
            return
        }
        var lifecycleListeners = mActivityLifecycleListeners[activity]
        if (lifecycleListeners == null) {
            lifecycleListeners = ArrayList()
        }
        lifecycleListeners.add(lifecycleListener)
        mActivityLifecycleListeners[activity] = lifecycleListeners
        //recursive字段设置true，将自动为所有子fragmentManager注册这个回调
        activity.supportFragmentManager.registerFragmentLifecycleCallbacks(this, true)
        info("add fragment lifecycle : " + activity.javaClass.simpleName)
    }

    fun unregisterFragmentLifecycleListener(
        activity: FragmentActivity?,
        lifecycleListener: FragmentLifecycleListener?
    ): Boolean {
        if (activity == null || lifecycleListener == null) {
            return false
        }
        synchronized(mActivityLifecycleListeners) {
            val lifecycleListeners = mActivityLifecycleListeners[activity] ?: return false
            val result = lifecycleListeners.remove(lifecycleListener)
            if (lifecycleListeners.size == 0) {
                mActivityLifecycleListeners.remove(activity)
            }
            activity.supportFragmentManager.unregisterFragmentLifecycleCallbacks(this)
            info("remove fragment lifecycle : " + activity.javaClass.simpleName)
            return result
        }
    }


    override fun onFragmentAttached(
        fm: FragmentManager,
        f: Fragment, context: Context
    ) {
        if (f.activity == null){
            return
        }
        val lifecycleListeners: List<FragmentLifecycleListener>? =
            mActivityLifecycleListeners[f.activity]
        if (lifecycleListeners == null || lifecycleListeners.isEmpty()) {
            return
        }
        val callbacks: List<FragmentLifecycleListener> = ArrayList(lifecycleListeners)
        for (i in callbacks.indices) {
            callbacks[i].onFragmentAttached(fm, f, context)
        }
    }

    override fun onFragmentCreated(fm: FragmentManager, f: Fragment, savedInstanceState: Bundle?) {
        if (f.activity == null){
            return
        }
        val lifecycleListeners: List<FragmentLifecycleListener>? =
            mActivityLifecycleListeners[f.activity]
        if (lifecycleListeners == null || lifecycleListeners.isEmpty()) {
            return
        }
        val callbacks: List<FragmentLifecycleListener> = ArrayList(lifecycleListeners)
        for (i in callbacks.indices) {
            callbacks[i].onFragmentCreated(fm, f, savedInstanceState)
        }
    }

    override fun onFragmentStarted(fm: FragmentManager, f: Fragment) {
        if (f.activity == null){
            return
        }
        val lifecycleListeners: List<FragmentLifecycleListener>? =
            mActivityLifecycleListeners[f.activity]
        if (lifecycleListeners == null || lifecycleListeners.isEmpty()) {
            return
        }
        val callbacks: List<FragmentLifecycleListener> = ArrayList(lifecycleListeners)
        for (i in callbacks.indices) {
            callbacks[i].onFragmentStarted(fm, f)
        }
    }

    override fun onFragmentResumed(fm: FragmentManager, f: Fragment) {
        if (f.activity == null){
            return
        }
        val lifecycleListeners: List<FragmentLifecycleListener>? =
            mActivityLifecycleListeners[f.activity]
        if (lifecycleListeners == null || lifecycleListeners.isEmpty()) {
            return
        }
        val callbacks: List<FragmentLifecycleListener> = ArrayList(lifecycleListeners)
        for (i in callbacks.indices) {
            callbacks[i].onFragmentResumed(fm, f)
        }
    }

    override fun onFragmentPaused(fm: FragmentManager, f: Fragment) {
        if (f.activity == null){
            return
        }
        val lifecycleListeners: List<FragmentLifecycleListener>? =
            mActivityLifecycleListeners[f.activity]
        if (lifecycleListeners == null || lifecycleListeners.isEmpty()) {
            return
        }
        val callbacks: List<FragmentLifecycleListener> = ArrayList(lifecycleListeners)
        for (i in callbacks.indices) {
            callbacks[i].onFragmentPaused(fm, f)
        }
    }

    override fun onFragmentStopped(fm: FragmentManager, f: Fragment) {
        if (f.activity == null){
            return
        }
        val lifecycleListeners: List<FragmentLifecycleListener>? =
            mActivityLifecycleListeners[f.activity]
        if (lifecycleListeners == null || lifecycleListeners.isEmpty()) {
            return
        }
        val callbacks: List<FragmentLifecycleListener> = ArrayList(lifecycleListeners)
        for (i in callbacks.indices) {
            callbacks[i].onFragmentStopped(fm, f)
        }
    }

    override fun onFragmentDestroyed(fm: FragmentManager, f: Fragment) {
        if (f.activity == null){
            return
        }
        val lifecycleListeners: List<FragmentLifecycleListener>? =
            mActivityLifecycleListeners[f.activity]
        if (lifecycleListeners == null || lifecycleListeners.isEmpty()) {
            return
        }
        val callbacks: List<FragmentLifecycleListener> = ArrayList(lifecycleListeners)
        for (i in callbacks.indices) {
            callbacks[i].onFragmentDestroyed(fm, f)
        }
    }

    override fun onFragmentDetached(fm: FragmentManager, f: Fragment) {
        if (f.activity == null){
            return
        }
        val lifecycleListeners: List<FragmentLifecycleListener>? =
            mActivityLifecycleListeners[f.activity]
        if (lifecycleListeners == null || lifecycleListeners.isEmpty()) {
            return
        }
        val callbacks: List<FragmentLifecycleListener> = ArrayList(lifecycleListeners)
        for (i in callbacks.indices) {
            callbacks[i].onFragmentDetached(fm, f)
        }
    }

    private fun info(s: String) {
        Log.i(com.yc.fragmentmanager.FragmentManager.TAG, s)
    }
}